### PR TITLE
Add SensitivityAtSpecificity metric

### PR DESCRIPTION
**Pull request by oguz-hanoglu**
## What does this PR do?

Fixes the remaining part of #971
It basically follows the same code structure with this [pull request](https://github.com/Lightning-AI/torchmetrics/pull/1432) and make the necessary modifications.
Repos like [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/SensitivityAtSpecificity) include both SensitivityAtSpecificity and SpecificityAtSensitivity, whereas Torchmetrics is missing SensitivityAtSpecificity. 

The github discussion at #971 ends with _"Feel free to implement the other variant SensitivityAtSpecificity if you guys are interested."_

<details>
  <summary>Before submitting</summary>

- [X] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section? (OH: no such a section, is this a test?)
- [ ] Did you make sure to **update the docs**?
- [X] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>
 SpecificityAtSensitivity is already included in TorchMetrics. This PR introduces the SensitivityAtSpecificity Metric.

</details>

## Did you have fun?

[X] Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2217.org.readthedocs.build/en/2217/

### CODE REVIEW

**Review by SkafteNicki — APPROVED** [review]

File: src/torchmetrics/classification/sensitivity_specificity.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        thresholds:
+            Can be one of:
+
+            - If set to `None`, will use a non-binned approach where thresholds are dynamically calculated from
+              all the data. Most accurate but also most memory consuming approach.
+            - If set to an `int` (larger than 1), will use that number of thresholds linearly spaced from
+              0 to 1 as bins for the calculation.
+            - If set to an `list` of floats, will use the indicated thresholds in the list as bins for the calculation
+            - If set to an 1d `tensor` of floats, will use the indicated thresholds in the tensor as
+              bins for the calculation.
```
Comment (Borda) [inline src/torchmetrics/classification/sensitivity_specificity.py:79]:
```suggestion
            - ``None``, will use a non-binned approach where thresholds are dynamically calculated from
              all the data. It is the most accurate but also the most memory-consuming approach.
            - ``int`` (larger than 1), will use that number of thresholds linearly spaced from
              0 to 1 as bins for the calculation.
            - ``list`` of floats, will use the indicated thresholds in the list as bins for the calculation
            - 1d ``tensor`` of floats, will use the indicated thresholds in the tensor as
              bins for the calculation.
```

File: src/torchmetrics/classification/sensitivity_specificity.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
```
Comment (Borda) [inline src/torchmetrics/classification/sensitivity_specificity.py:168]:
```suggestion
            - ``None``, will use a non-binned approach where thresholds are dynamically calculated from
              all the data. It is the most accurate but also the most memory-consuming approach.
            - ``int`` (larger than 1), will use that number of thresholds linearly spaced from
              0 to 1 as bins for the calculation.
            - ``list`` of floats, will use the indicated thresholds in the list as bins for the calculation
            - 1d ``tensor`` of floats, will use the indicated thresholds in the tensor as
              bins for the calculation.
```

File: src/torchmetrics/classification/sensitivity_specificity.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
```
Comment (Borda) [inline src/torchmetrics/classification/sensitivity_specificity.py:266]:
```suggestion
            - ``None``, will use a non-binned approach where thresholds are dynamically calculated from
              all the data. It is the most accurate but also the most memory-consuming approach.
            - ``int`` (larger than 1), will use that number of thresholds linearly spaced from
              0 to 1 as bins for the calculation.
            - ``list`` of floats, will use the indicated thresholds in the list as bins for the calculation
            - 1d ``tensor`` of floats, will use the indicated thresholds in the tensor as
              bins for the calculation.
```

File: src/torchmetrics/functional/classification/sensitivity_specificity.py [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+            Can be one of:
+
+            - If set to `None`, will use a non-binned approach where thresholds are dynamically calculated from
+              all the data. Most accurate but also most memory consuming approach.
+            - If set to an `int` (larger than 1), will use that number of thresholds linearly spaced from
+              0 to 1 as bins for the calculation.
+            - If set to an `list` of floats, will use the indicated thresholds in the list as bins for the calculation
+            - If set to an 1d `tensor` of floats, will use the indicated thresholds in the tensor as
+              bins for the calculation.
+
```
Comment (Borda) [inline src/torchmetrics/functional/classification/sensitivity_specificity.py:139]:
```suggestion
            - ``None``, will use a non-binned approach where thresholds are dynamically calculated from
              all the data. It is the most accurate but also the most memory-consuming approach.
            - ``int`` (larger than 1), will use that number of thresholds linearly spaced from
              0 to 1 as bins for the calculation.
            - ``list`` of floats, will use the indicated thresholds in the list as bins for the calculation
            - 1d ``tensor`` of floats, will use the indicated thresholds in the tensor as
              bins for the calculation.

```

File: src/torchmetrics/functional/classification/sensitivity_specificity.py [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+        thresholds:
+            Can be one of:
+
+            - If set to `None`, will use a non-binned approach where thresholds are dynamically calculated from
+              all the data. Most accurate but also most memory consuming approach.
+            - If set to an `int` (larger than 1), will use that number of thresholds linearly spaced from
+              0 to 1 as bins for the calculation.
+            - If set to an `list` of floats, will use the indicated thresholds in the list as bins for the calculation
+            - If set to an 1d `tensor` of floats, will use the indicated thresholds in the tensor as
+              bins for the calculation.
```
Comment (Borda) [inline src/torchmetrics/functional/classification/sensitivity_specificity.py:249]:
```suggestion
            - ``None``, will use a non-binned approach where thresholds are dynamically calculated from
              all the data. It is the most accurate but also the most memory-consuming approach.
            - ``int`` (larger than 1), will use that number of thresholds linearly spaced from
              0 to 1 as bins for the calculation.
            - ``list`` of floats, will use the indicated thresholds in the list as bins for the calculation
            - 1d ``tensor`` of floats, will use the indicated thresholds in the tensor as
              bins for the calculation.
```

File: src/torchmetrics/functional/classification/sensitivity_specificity.py [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
```
Comment (Borda) [inline src/torchmetrics/functional/classification/sensitivity_specificity.py:366]:
```suggestion
            - ``None``, will use a non-binned approach where thresholds are dynamically calculated from
              all the data. It is the most accurate but also the most memory-consuming approach.
            - ``int`` (larger than 1), will use that number of thresholds linearly spaced from
              0 to 1 as bins for the calculation.
            - ``list`` of floats, will use the indicated thresholds in the list as bins for the calculation
            - 1d ``tensor`` of floats, will use the indicated thresholds in the tensor as
              bins for the calculation.
```

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/classification/sensitivity_specificity.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+        validate_args: bool indicating if input arguments and tensors should be validated for correctness.
+            Set to ``False`` for faster computations.
+        kwargs: Additional keyword arguments, see :ref:`Metric kwargs` for more info.
+
+    Returns:
+        (tuple): a tuple of either 2 tensors or 2 lists containing
+
+        - sensitivity: an 1d tensor of size (n_classes, ) with the maximum sensitivity for the given
+            specificity level per class
+        - thresholds: an 1d tensor of size (n_classes, ) with the corresponding threshold level per class
```
Comment (Borda) [inline src/torchmetrics/classification/sensitivity_specificity.py:277]:
```suggestion
        - sensitivity: an 1d tensor of size ``(n_classes, )`` with the maximum sensitivity for the given
            specificity level per class
        - thresholds: an 1d tensor of size ``(n_classes, )`` with the corresponding threshold level per class
```

**Comment by Borda** [comment]
@oguz-hanoglu @SkafteNicki seems like a real issue with GPU and old version of pyTorch

**Comment by Borda** [comment]
@oguz-hanoglu @SkafteNicki seems to be constantly failing on PT 1.10
see: https://dev.azure.com/Lightning-AI/Metrics/_build/results?buildId=189940&view=logs&jobId=95985737-7c02-5ecd-7847-688de4203f7e&j=95985737-7c02-5ecd-7847-688de4203f7e&t=30568a28-ed27-55ff-e099-f25174540c05

**Comment by Borda** [comment]
we can also allow this method only for newer PT
